### PR TITLE
[clang][dataflow] Add a lattice to help cache const accessor methods

### PR DESCRIPTION
**Pull request by jvoung**
By caching const accessor methods we can sometimes treat method call results as stable (e.g., for issue https://github.com/llvm/llvm-project/issues/58510). Users can clear the cache when a non-const method is called that may modify the state of an object.
This is represented as mixin.

It  will be used in a follow on patch to change bugprone-unchecked-optional-access's lattice from NoopLattice to CachedConstAccessorsLattice<NoopLattice>, along with some additional transfer functions.

### CODE REVIEW

File: clang/unittests/Analysis/FlowSensitive/CachedConstAccessorsLatticeTest.cpp [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+  Environment Env{DACtx};
+};
+
+// Basic test AST with two const methods (return a value, and return a ref).
+struct CommonTestInputs {
+  CommonTestInputs()
+      : AST(R"cpp(
+    struct S {
+      int *valProperty() const;
+      int &refProperty() const;
```
Comment (Xazax-hun) [inline clang/unittests/Analysis/FlowSensitive/CachedConstAccessorsLatticeTest.cpp:69]: I'd like to see a test with a method returning a record.

**Review by Xazax-hun — COMMENTED** [review]
Overall looks good to me but also wait for a review from the code owners.

File: clang/include/clang/Analysis/FlowSensitive/CachedConstAccessorsLattice.h [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+///     s.clear();
+///     use(s.getFoo().value()); // unsafe (invalidate cache for s)
+///   }
+/// }
+template <typename Base> class CachedConstAccessorsLattice : public Base {
+public:
+  using Base::Base; // inherit all constructors
+
+  /// Creates or returns a previously created `Value` associated with a const
+  ///  method call `obj.getFoo()` where `RecordLoc` is the
```
Comment (ymand) [inline clang/include/clang/Analysis/FlowSensitive/CachedConstAccessorsLattice.h:53]:
```suggestion
  /// method call `obj.getFoo()` where `RecordLoc` is the
```

File: clang/include/clang/Analysis/FlowSensitive/CachedConstAccessorsLattice.h [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+  /// Returns nullptr if unable to find or create a value.
+  ///
+  /// Requirements:
+  ///
+  ///  - `CE` should return a value (not a reference or record type)
+  Value *
+  getOrCreateConstMethodReturnValue(const RecordStorageLocation &RecordLoc,
+                                    const CallExpr *CE, Environment &Env);
+
+  /// Creates or returns a previously created `StorageLocation` associated a
```
Comment (ymand) [inline clang/include/clang/Analysis/FlowSensitive/CachedConstAccessorsLattice.h:64]:
```suggestion
  /// Creates or returns a previously created `StorageLocation` associated with a
```

File: clang/unittests/Analysis/FlowSensitive/CachedConstAccessorsLatticeTest.cpp [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    };
+    void target() {
+      S s;
+      s.valProperty();
+      S s2;
```
Comment (ymand) [inline clang/unittests/Analysis/FlowSensitive/CachedConstAccessorsLatticeTest.cpp:74]: Can we also test: 2 objects, same function called, distinct results? That would catch bad keying on the object (for example).

**Review by ymand — APPROVED** [review]
modulo the requests for tests. This looks great!

File: clang/unittests/Analysis/FlowSensitive/CachedConstAccessorsLatticeTest.cpp (hunk 1 above)
Comment (jvoung) [reply clang/unittests/Analysis/FlowSensitive/CachedConstAccessorsLatticeTest.cpp:72]: Thanks! Added a test "SameStructValBeforeClearOrDiffAfterClear"

File: clang/unittests/Analysis/FlowSensitive/CachedConstAccessorsLatticeTest.cpp (hunk 4 above)
Comment (jvoung) [reply clang/unittests/Analysis/FlowSensitive/CachedConstAccessorsLatticeTest.cpp:77]: Thanks! Added "DifferentValsFromDifferentLocs"

File: clang/include/clang/Analysis/FlowSensitive/CachedConstAccessorsLattice.h (hunk 2 above)
Comment (jvoung) [reply clang/include/clang/Analysis/FlowSensitive/CachedConstAccessorsLattice.h]: Fixed!

File: clang/include/clang/Analysis/FlowSensitive/CachedConstAccessorsLattice.h (hunk 3 above)
Comment (jvoung) [reply clang/include/clang/Analysis/FlowSensitive/CachedConstAccessorsLattice.h]: Fixed!

**Review by ymand — APPROVED** [review]

File: clang/include/clang/Analysis/FlowSensitive/CachedConstAccessorsLattice.h [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+  // Maps a record storage location and const method to the record storage
+  // location to return from that const method.
+  using ConstMethodReturnStorageLocationsType = llvm::SmallDenseMap<
+      const RecordStorageLocation *,
+      llvm::SmallDenseMap<const FunctionDecl *, StorageLocation *>>;
+  ConstMethodReturnStorageLocationsType ConstMethodReturnStorageLocations;
+};
+
+namespace internal {
```
Comment (nico) [inline clang/include/clang/Analysis/FlowSensitive/CachedConstAccessorsLattice.h:109]: This breaks building on Windows (arguably by exposing an existing bug):

http://45.33.8.238/win/95986/step_3.txt (scroll to the very bottom):

```
../../clang/include\clang/Analysis/FlowSensitive/CachedConstAccessorsLattice.h(157,29): error: reference to 'internal' is ambiguous
  157 |   ConstMethodReturnValues = internal::joinConstMethodMap<Value>(
      |                             ^
../../clang/lib/Analysis/FlowSensitive/Models/UncheckedOptionalAccessModel.cpp(1083,31): note: in instantiation of member function 'clang::dataflow::DataflowAnalysis<clang::dataflow::UncheckedOptionalAccessModel, clang::dataflow::CachedConstAccessorsLattice<clang::dataflow::NoopLattice>>::joinTypeErased' requested here
 1083 | UncheckedOptionalAccessModel::UncheckedOptionalAccessModel(ASTContext &Ctx,
      |                               ^
../../clang/include\clang/Analysis/FlowSensitive/CachedConstAccessorsLattice.h(109,11): note: candidate found by name lookup is 'clang::dataflow::internal'
  109 | namespace internal {
      |           ^
../../clang/lib/Analysis/FlowSensitive/Models/UncheckedOptionalAccessModel.cpp(205,1): note: candidate found by name lookup is 'clang::dataflow::(anonymous namespace)::internal'
  205 | AST_MATCHER_P(CXXMemberCallExpr, publicReceiverType,
      | ^
../../clang/include\clang/ASTMatchers/ASTMatchersMacros.h(131,3): note: expanded from macro 'AST_MATCHER_P'
  131 |   AST_MATCHER_P_OVERLOAD(Type, DefineMatcher, ParamType, Param, 0)
      |   ^
../../clang/include\clang/ASTMatchers/ASTMatchersMacros.h(135,13): note: expanded from macro 'AST_MATCHER_P_OVERLOAD'
  135 |   namespace internal {                                                         \
      |             ^
```

Please take a look and revert for now if it takes a while to fix.

Comment (jvoung) [reply clang/include/clang/Analysis/FlowSensitive/CachedConstAccessorsLattice.h:109]: Ah -- Taking a look!

Comment (nico) [reply clang/include/clang/Analysis/FlowSensitive/CachedConstAccessorsLattice.h:109]: (Here's a build that needs less scrolling to find the diag: http://45.33.8.238/win/95994/step_3.txt)

Comment (jvoung) [reply clang/include/clang/Analysis/FlowSensitive/CachedConstAccessorsLattice.h:109]: Thanks! Created https://github.com/llvm/llvm-project/pull/113601

Though is there a way to validate this, similar to this clang-cl compile?

**Pull request merged**